internal/configs/resource.go

```diff
@@ -813,59 +929,29 @@ func (m *ProviderConfigRefMapping) decodeStaticAlias(eval *StaticEvaluator, forE
 		})
 	}
 
-	// TODO/Oleksandr: Is static identifier valid if we evaluate for_each for alias?
-	instances, instancesDiags := eval.Evaluate(forEach, aliasStaticID)
+	instances, instancesDiags := eval.Evaluate(instanceExpr, aliasStaticID)
 	diags = diags.Extend(instancesDiags)
 	if instancesDiags.HasErrors() {
 		return diags
 	}
 
-	if !instances.CanIterateElements() {
-		panic("decodeStaticAlias must be called with for each expression: wrong expression?")
-	}
-
-	aliases := make(map[addrs.InstanceKey]string)
-
-	instanceIter := instances.ElementIterator()
-	for instanceIter.Next() {
-		k, v := instanceIter.Element()
-
-		instanceEvalCtx := aliasEvalCtx.NewChild()
-		instanceEvalCtx.Variables = map[string]cty.Value{
-			"each": cty.ObjectVal(map[string]cty.Value{
-				"key":   k,
-				"value": v,
-			}),
-		}
+	var aliases map[addrs.InstanceKey]string
+	var aliasDiags hcl.Diagnostics
 
-		aliasVal, aliasDiags := m.Alias.Value(instanceEvalCtx)
-		diags = diags.Extend(aliasDiags)
-		if aliasDiags.HasErrors() {
-			continue
-		}
-
-		if !aliasVal.Type().Equals(cty.String) {
-			// TODO/Oleksandr: add friendly error
-			return diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid alias",
-				Detail:   "",
-				Subject:  m.Alias.Range().Ptr(),
-			})
-		}
-
-		instanceKey, err := addrs.ParseInstanceKey(k)
-		if err != nil {
-			panic("decodeStaticAlias failed to parse for each instance keys: wrong expression?")
-		}
-
-		aliases[instanceKey] = aliasVal.AsString()
+	// Default case is handled earlier to provider a NoKey alias.
+	switch {
+	case hasEachRefInAlias:
+		aliases, aliasDiags = generateForEachAliases(instances, aliasEvalCtx, m.Alias)
+	case hasCountRefInAlias:
+		aliases, aliasDiags = generateCountAliases(instances, aliasEvalCtx, m.Alias)
 	}
 
-	if !diags.HasErrors() {
-		m.Aliases = aliases
+	diags = diags.Extend(aliasDiags)
+	if aliasDiags.HasErrors() {
+		return diags
 	}
 
+	m.Aliases = aliases
```

Comment:
We should at some point validate that these aliases are valid identifiers, either in this PR or a followup.

Reply:
Previously, we didn't have such a validation. Also, I think we need just to check if provider exists - it will cover everything else. What do you think?